watertap/tools/parameter_sweep/parameter_sweep_class.py

```diff
@@ -0,0 +1,1066 @@
+        reinitialize_kwargs=None,
+        reinitialize_before_sweep=False,
+        probe_function=None,
+        num_samples=None,
+        seed=None,
+    ):
+
+        # Convert sweep_params to LinearSamples
+        sweep_params, sampling_type = self._process_sweep_params(sweep_params)
+
+        # Set the seed before sampling
+        np.random.seed(seed)
+
+        # Enumerate/Sample the parameter space
+        global_values = self._build_combinations(
+            sweep_params, sampling_type, num_samples
+        )
+
+        # divide the workload between processors
+        local_values = self._divide_combinations(global_values)
+        local_num_cases = np.shape(local_values)[0]
+
+        # Set up optimize_kwargs
+        if optimize_kwargs is None:
+            optimize_kwargs = dict()
+        # Set up reinitialize_kwargs
+        if reinitialize_kwargs is None:
+            reinitialize_kwargs = dict()
+
+        # Do the Loop
+        local_results_dict = self._do_param_sweep(
+            model,
+            sweep_params,
+            outputs,
+            local_values,
+            optimize_function,
+            optimize_kwargs,
+            reinitialize_function,
+            reinitialize_kwargs,
+            reinitialize_before_sweep,
+            probe_function,
+        )
+
+        # Aggregate results on Master
+        global_results_dict, global_results_arr = self._aggregate_local_results(
+            global_values, local_results_dict, num_samples, local_num_cases
+        )
+
+        # Save to file
+        global_save_data = self.writer.save_results(
+            sweep_params,
+            local_values,
+            global_values,
+            local_results_dict,
+            global_results_dict,
+            global_results_arr,
+        )
+
+        return global_save_data, global_results_dict
+
+
+class RecursiveParameterSweep(_ParameterSweepBase):
+    def __init__(
+        self,
+        csv_results_file_name=None,
+        h5_results_file_name=None,
+        debugging_data_dir=None,
+        interpolate_nan_outputs=False,
+        *args,
+        **kwargs,
+    ):
+
+        # Initialize the base Class
+        _ParameterSweepBase.__init__(self)
+
+        # Initialize the writer
+        # if self.rank == 0:
+        #     if h5_results_file_name is None and csv_results_file_name is None:
+        #         warnings.warn(
+        #             "No results will be writen to disk as h5_results_file_name and csv_results_file_name are both None"
+        #         )
+        #         self.write_outputs = False
+        #     else:
+        #         self.write_outputs = True
+        self.writer = ParameterSweepWriter(
+            self.comm,
+            csv_results_file_name=csv_results_file_name,
+            h5_results_file_name=h5_results_file_name,
+            debugging_data_dir=debugging_data_dir,
+            interpolate_nan_outputs=interpolate_nan_outputs,
+        )
```

Comment:
It looks to me like both can go in `_ParameterSweepBase.__init__`

Reply:
Moved into a base class in a more recent commit.